File: CSharpPools/PoolsLibrary/Pool/BasicFunctionality/Item/PoolItem.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using PoolsLibrary.Controller;
using PoolsLibrary.ObjectActions;
using PoolsLibrary.ObjectUtilization;
using PoolsLibrary.Pool.BasicFunctionality.Storages;

namespace PoolsLibrary.Pool.BasicFunctionality.Item
{
    class PoolItem<TK, TV> : IDisposable
    {
        public IStorage<TV> AvailableObjects { get; set; }
        public int NotAvailableObjectsCount
        {
            get { return _notAvailableObjects.Count; }
        }
        public int AllObjectsCount
        {
            get { return NotAvailableObjectsCount + AvailableObjects.Count; }
        }

        private readonly PoolItemSettings<TK> _settings;
        private readonly HashSet<TV> _notAvailableObjects = new HashSet<TV>();
        private readonly HashSet<TV> _objectsToKill = new HashSet<TV>();
        private readonly object _syncObject = new object();

        private readonly IPoolObjectActions<TV> _objectActions;
        private readonly IObjectUtilizer<TK, TV> _objectUtilizer;

        public PoolItem(PoolItemSettings<TK> settings,
                        IStorage<TV> availableObjectsStorage,
                        IPoolObjectActions<TV> objectActions,
                        IObjectUtilizer<TK, TV> objectUtilizer)
        {
            CheckConstructorArguments(settings, availableObjectsStorage, objectActions, objectUtilizer);

            _settings = settings;
            AvailableObjects = availableObjectsStorage;
            _objectActions = objectActions;
            _objectUtilizer = objectUtilizer;
        }

        public bool TryObtain(out TV outPoolObject, Func<TK, TV> createDelegateIfNoObjectIsAvailable)
        {
            SharedEnvironment.Log.DebugFormat("Obtaining has started. Delegate='{0}'", createDelegateIfNoObjectIsAvailable);

            lock (_syncObject)
            {
                SharedEnvironment.Log.DebugFormat("Obtaining has acquired lock. {0} total, {1} idle",
                                                   AllObjectsCount, AvailableObjects.Count);

                if (!TryProvideExistingObject(out outPoolObject) &&
                    !TryProvideNewObject(createDelegateIfNoObjectIsAvailable, out outPoolObject))
                {
                    outPoolObject = default(TV);
                    return false;
                }
                SharedEnvironment.Log.DebugFormat("Obtaining has got object. {0} total, {1} idle",
                                                   AllObjectsCount, AvailableObjects.Count);

                if (_settings.MarkObtainedObjectAsNotAvailable)
                    _notAvailableObjects.Add(outPoolObject);
                else
                    AvailableObjects.Add(outPoolObject);
                SharedEnvironment.Log.DebugFormat("Obtaining has marked object. {0} total, {1} idle",
                                                   AllObjectsCount, AvailableObjects.Count);
            }

            SharedEnvironment.Log.DebugFormat("Obtaining has released lock");
            return true;
        }

        public void Release(TV objectToUnmark)
        {
            SharedEnvironment.Log.Debug("Releasing has started");

            if (!_settings.MarkObtainedObjectAsNotAvailable)
            {
                var markingIsOffMessage = "Operation of marking object as available is invalid because marking was ordered to be off";
                throw new InvalidPoolOperationException<TK, TV>(objectToUnmark, _settings.Key, markingIsOffMessage);
            }

            lock (_syncObject)
            {
                SharedEnvironment.Log.DebugFormat("Releasing has acquired lock. {0} total, {1} idle",
                                                   AllObjectsCount, AvailableObjects.Count);

                if (_notAvailableObjects.Remove(objectToUnmark))
                {
                    SharedEnvironment.Log.Debug("Object has been removed from busy objects collection");
                    if (!KilledBecauseItIsBad(objectToUnmark))
                    {
                        AvailableObjects.Add(objectToUnmark);
                        SharedEnvironment.Log.Debug("Object has been added to free objects collection");
                    }

                    SharedEnvironment.Log.DebugFormat("Releasing is about to release lock. {0} total, {1} idle",
                                                       AllObjectsCount, AvailableObjects.Count);
                    return;
                }

                var invalidUnmarkOperationMessage = AvailableObjects.Contains(objectToUnmark)
                    ? "Marking object as available has been declined because it's currently available. Object should be marked as not available first"
                    : "Marking object as available has been declined because this object wasn't created by pool, it's a stranger";
                throw new InvalidPoolOperationException<TK, TV>(objectToUnmark, _settings.Key, invalidUnmarkOperationMessage);
            }
        }

        public void MarkObjectForKilling(TV toKill)
        {
            lock (_syncObject)
            {
                if (_notAvailableObjects.Contains(toKill) || AvailableObjects.Contains(toKill))
                    _objectsToKill.Add(toKill);
            }
        }

        public void Dispose()
        {
            SharedEnvironment.Log.DebugFormat("Disposing has started. {0} total, {1} idle",
                                               AllObjectsCount, AvailableObjects.Count);

            while (AvailableObjects.Count > 0)
                KillObject(AvailableObjects.Remove());
            SharedEnvironment.Log.Debug("Idle objects have been killed");

            foreach (var notAvailableObject in _notAvailableObjects)
                KillObject(notAvailableObject);
            SharedEnvironment.Log.Debug("Busy objects have been killed");
        }

        private bool TryProvideExistingObject(out TV outPoolObject)
        {
            while (AvailableObjects.Count > 0)
            {
                var availableObject = AvailableObjects.Remove();

                if (KilledBecauseItIsBad(availableObject))
                    continue;

                outPoolObject = availableObject;
                return true;
            }

            outPoolObject = default(TV);
            return false;
        }

        private bool TryProvideNewObject(Func<TK, TV> createDelegate, out TV outPoolObject)
        {
            if (AllObjectsCount == _settings.MaxObjectsCount)
                if (_settings.ThrowIfCantCreateNewBecauseOfReachedLimit)
                {
                    var limitReachedMessage = string.Format("Object with key='{0}' wasn't created because " +
                                                            "max objects count {1} is already reached",
                                                            _settings.Key, _settings.MaxObjectsCount);
                    throw new ObjectsMaxCountReachedException<TK>(_settings.Key, _settings.MaxObjectsCount, limitReachedMessage);
                }
                else
                {
                    outPoolObject = default(TV);
                    return false;
                }

            if (createDelegate == null)
            {
                outPoolObject = default(TV);
                return false;
            }

            TV newlyCreatedObject;
            try
            {
                newlyCreatedObject = createDelegate(_settings.Key);
            }
            catch (Exception e)
            {
                var creationThrewMessage = string.Format("Creation object with key='{0}' failed. Look at inner exception for details",
                                                         _settings.Key);
                throw new ObjectCreationFailedException<TK, TV>(_settings.Key, createDelegate, creationThrewMessage, e);
            }

            if (_objectActions.IsValid(newlyCreatedObject))
            {
                outPoolObject = newlyCreatedObject;
                return true;
            }

            _objectActions.Dispose(newlyCreatedObject);
            var invalidObjectCreatedMessage = string.Format("Provided delegate created invalid object with key='{0}'",
                                                            _settings.Key);
            throw new ObjectCreationFailedException<TK, TV>(_settings.Key, createDelegate, invalidObjectCreatedMessage);
        }

        private bool KilledBecauseItIsBad(TV toCheck)
        {
            if (!_objectsToKill.Remove(toCheck) && _objectActions.IsValid(toCheck))
                return false;

            KillObject(toCheck);
            return true;
        }

        private void KillObject(TV toKill)
        {
            _objectActions.Dispose(toKill);
            _objectUtilizer.Utilize(_settings.Key, toKill, this);
        }

        private void CheckConstructorArguments(PoolItemSettings<TK> settings,
                                       IStorage<TV> availableObjectsStorage,
                                       IPoolObjectActions<TV> objectActions,
                                       IObjectUtilizer<TK, TV> objectUtilizer)
        {
            if (settings == null)
                throw new ArgumentNullException("settings");

            if (settings.Key == null)
                throw new ArgumentException("PoolItemSettings should contain key", "settings.Key");

            if (availableObjectsStorage == null)
                throw new ArgumentNullException("availableObjectsStorage");

            if (objectActions == null)
                throw new ArgumentNullException("objectActions");

            if (objectUtilizer == null)
                throw new ArgumentNullException("objectUtilizer");
        }
    }
}

File: JavaPools/PoolsLibrary/src/main/java/ru/urfu/javapools/poolslibrary/pool/basicfunctionality/item/PoolItem.java
package ru.urfu.javapools.poolslibrary.pool.basicfunctionality.item;

import java.util.Collection;
import java.util.HashSet;
import java.util.Iterator;

import ru.urfu.javapools.poolslibrary.controller.InvalidPoolOperationException;
import ru.urfu.javapools.poolslibrary.controller.PoolObjectObtainingInterruptedException;
import ru.urfu.javapools.poolslibrary.controller.PoolObjectReleasingInterruptedException;
import ru.urfu.javapools.poolslibrary.function.FunctionThatMayThrow;
import ru.urfu.javapools.poolslibrary.lang.MyReference;
import ru.urfu.javapools.poolslibrary.locks.FairLockWithPrecedence;
import ru.urfu.javapools.poolslibrary.objectactions.IPoolObjectActions;

public class PoolItem<TK, TV> implements AutoCloseable {

	private final PoolItemSettings<TK> _settings;
	private final Collection<TV> _availableObjects;
	private final HashSet<TV> _notAvailableObjects = new HashSet<TV>();
	private final HashSet<TV> _objectsToKill = new HashSet<TV>();

	private final FairLockWithPrecedence _obtainReleaseLock = new FairLockWithPrecedence();
	
	private final IPoolObjectActions<TV> _objectActions;

	private static final String _MESSAGE_MARKING_IS_OFF = "Operation of marking object as available is invalid because marking was ordered to be off";
	private static final String _MESSAGE_UNMARKING_DECLINED = "Marking object as available has been declined";
	private static final String _MESSAGE_REASON_CURRENTLY_AVAILABLE = "because it's currently available. Object should be marked as not available first";
	private static final String _MESSAGE_REASON_STRANGER = "because this object wasn't created by pool, it's a stranger";

	public PoolItem(PoolItemSettings<TK> settings,
					Collection<TV> availableObjectsStorage,
					IPoolObjectActions<TV> objectActions) {
		
		_settings = settings;
		_availableObjects = availableObjectsStorage;
		_objectActions = objectActions;
	}
	
	public TV obtain(FunctionThatMayThrow<TK, TV> createDelegateIfNoObjectIsAvailable)
		throws NoAvailableObjectException, ObjectsMaxCountReachedException,
			   ObjectCreationFailedException, PoolObjectObtainingInterruptedException {
		
		try {
			_obtainReleaseLock.lock();
		} catch (InterruptedException e) {
			throw new PoolObjectObtainingInterruptedException(_settings.getKey(), e);
		}
	
		try {
			if (_notAvailableObjects.size() == _settings.getMaxObjectsCount())
				throw new ObjectsMaxCountReachedException(_settings.getKey(), _settings.getMaxObjectsCount());

			MyReference<TV> out = new MyReference<TV>(null);
			if (tryGetExistingAvailableObject(out) || tryCreateNewObject(out,createDelegateIfNoObjectIsAvailable)) {			
				markBeforeGiving(out.get());
				return out.get();
			}

			throw new NoAvailableObjectException(_settings.getKey());
		} finally {
			_obtainReleaseLock.unlock();
		}
	}

	public void release(TV objectToUnmark) throws InvalidPoolOperationException, PoolObjectReleasingInterruptedException {

		if (!_settings.getMarkObtainedObjectAsNotAvailable())
			throw new InvalidPoolOperationException(_settings.getKey(), objectToUnmark, _MESSAGE_MARKING_IS_OFF);

		try {
			_obtainReleaseLock.lockImportant();
		} catch (InterruptedException e) {
			throw new PoolObjectReleasingInterruptedException(_settings.getKey(), objectToUnmark, e);
		}
		
		try {
			if (!_notAvailableObjects.remove(objectToUnmark)) {
				String invalidUnmarkOperationMessage = _MESSAGE_UNMARKING_DECLINED + " " + (_availableObjects.contains(objectToUnmark)
																							? _MESSAGE_REASON_CURRENTLY_AVAILABLE
																							: _MESSAGE_REASON_STRANGER);
				throw new InvalidPoolOperationException(_settings.getKey(), objectToUnmark, invalidUnmarkOperationMessage);
			}

			if (!closeIfBad(objectToUnmark))
				_availableObjects.add(objectToUnmark);
		} finally {
			_obtainReleaseLock.unlock();
		}
	}

	public void markObjectForKilling(TV toKill) {

		if (_availableObjects.remove(toKill))
			return;
		
		if (_notAvailableObjects.contains(toKill))
			_objectsToKill.add(toKill);
	}

	@Override
	public void close() throws Exception {
		
		Iterator<TV> avaliableObjectsIterator = _availableObjects.iterator();
		while (avaliableObjectsIterator.hasNext())
			_objectActions.close(avaliableObjectsIterator.next());

		for (TV notAvailableObject : _notAvailableObjects)
			_objectActions.close(notAvailableObject);
	}

	public int getAllObjectsCount() {
		return _availableObjects.size() + _notAvailableObjects.size();
	}

	private boolean tryGetExistingAvailableObject(MyReference<TV> out) {

		Iterator<TV> avaliableObjectIterator = _availableObjects.iterator();
		while (avaliableObjectIterator.hasNext()) {
			out.set(avaliableObjectIterator.next());
			avaliableObjectIterator.remove();
			if (!closeIfBad(out.get()))
				return true;
		}

		return false;
	}

	private boolean tryCreateNewObject(MyReference<TV> out, FunctionThatMayThrow<TK, TV> createDelegateIfNoObjectIsAvailable)
		throws ObjectCreationFailedException {
		
		if (createDelegateIfNoObjectIsAvailable == null)
			return false;

		try {
			out.set(createDelegateIfNoObjectIsAvailable.apply(_settings.getKey()));
			return true;
		} catch (Exception e) {
			throw new ObjectCreationFailedException(_settings.getKey(), createDelegateIfNoObjectIsAvailable, e);
		}
	}

	private boolean closeIfBad(TV object) {

		if (!_objectsToKill.remove(object) && _objectActions.isValid(object))
			return false;

		_objectActions.close(object);
		return true;
	}

	private void markBeforeGiving(TV toProvide) {

		if (_settings.getMarkObtainedObjectAsNotAvailable())
			_notAvailableObjects.add(toProvide);
		else
			_availableObjects.add(toProvide);
	}
}
